lightpath/controller.py

```diff
@@ -165,18 +166,46 @@ def load_beamline(self, endstation: str):
             # Find the paths from each source to the desired line
             for src in self.sources:
                 try:
-                    if nx.has_path(self.graph, src, branch):
-                        found_paths = nx.all_simple_paths(self.graph,
-                                                          source=src,
-                                                          target=branch)
-                        paths.extend(found_paths)
-                    else:
-                        logger.debug(f'No path between {src} and {branch}')
+                    path_exists: bool = nx.has_path(self.graph, src, branch)
                 except NodeNotFound:
                     logger.debug(f'Either source {src} or target {branch} '
                                  'not found.')
+                    path_exists = False
+
+                if path_exists:
+                    found_paths = nx.all_simple_paths(self.graph,
+                                                      source=src,
+                                                      target=branch)
+
+                    # Trim beamline based on z-positions
+                    if isinstance(end_branches, dict):
+                        new_paths = list()
+                        for found_path in found_paths:
+                            end_z = end_branches[branch]
+                            if end_z is not None:
+                                # Only filter internal devices
+                                truncated_path = [
+                                    name for name in found_path[1:-1]
+                                    if self.graph.nodes[name]['md'].res['z'] < end_z
```

Comment:
Does this mean we need to pair this with accurately setting end_z based on which devices each hutch cares about?

Reply:
Yep.  Currently this only matters for XPP, which ends halfway through the `L0` line.  We have already been defining these in lightpath, it's just that we used to split the path using the `BeamPath` object
https://github.com/pcdshub/lightpath/blob/22ec6aa7cce4c5133d3edf304c3eec28c24c5ccc/lightpath/config.py#L8-L16